token/services/interop/exchange/transaction.go

```diff
@@ -0,0 +1,342 @@
+			}
+			if hashFunc == 0 {
+				hashFunc = crypto.SHA256 // default hash function
+			}
+		}
+		boxed, ok = options.Attributes["exchange.hashEncoding"]
+		if ok {
+			hashEncoding, ok = boxed.(encoding.Encoding)
+			if !ok {
+				return nil, errors.Errorf("expected exchange.hashEncoding attribute to be Encoding, got [%T]", boxed)
+			}
+		}
+	}
+	script, preImage, err := t.recipientAsScript(sender, recipient, deadline, hash, hashFunc, hashEncoding)
+	if err != nil {
+		return nil, err
+	}
+	_, err = t.TokenRequest.Transfer(wallet, typ, []uint64{value}, []view.Identity{script}, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return preImage, nil
+}
+
+// Reclaim appends a reclaim (transfer) action to the token request of the transaction
+func (t *Transaction) Reclaim(wallet *token.OwnerWallet, tok *token2.UnspentToken) error {
+	// TODO: handle this properly
+	q, err := token2.ToQuantity(tok.Quantity, t.TokenRequest.TokenService.PublicParametersManager().Precision())
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert quantity [%s]", tok.Quantity)
+	}
+	owner, err := identity.UnmarshallRawOwner(tok.Owner.Raw)
+	if err != nil {
+		return err
+	}
+	if owner.Type != ScriptTypeExchange {
+		return errors.Errorf("invalid owner type, expected exchange script")
+	}
+	script := &Script{}
+	err = json.Unmarshal(owner.Identity, script)
+	if err != nil {
+		return errors.Errorf("failed to unmarshal RawOwner as an exchange script")
+	}
+
+	// Register the signer for the reclaim
+	sigService := t.TokenService().SigService()
+	signer, err := sigService.GetSigner(script.Sender)
+	if err != nil {
+		return err
+	}
+	verifier, err := sigService.OwnerVerifier(script.Sender)
+	if err != nil {
+		return err
+	}
+	logger.Debugf("registering signer for reclaim...")
+	if err := sigService.RegisterSigner(
+		tok.Owner.Raw,
+		signer,
+		verifier,
+	); err != nil {
+		return err
+	}
+
+	if err := view2.GetEndpointService(t.SP).Bind(script.Sender, tok.Owner.Raw); err != nil {
+		return err
+	}
+
+	return t.Transfer(wallet, tok.Type, []uint64{q.ToBigInt().Uint64()}, []view.Identity{script.Sender}, token.WithTokenIDs(tok.Id))
+}
+
+// Claim appends a claim (transfer) action to the token request of the transaction
+func (t *Transaction) Claim(wallet *token.OwnerWallet, tok *token2.UnspentToken, preImage []byte) error {
+	// TODO: handle this properly
+	q, err := token2.ToQuantity(tok.Quantity, t.TokenRequest.TokenService.PublicParametersManager().Precision())
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert quantity [%s]", tok.Quantity)
+	}
+
+	owner, err := identity.UnmarshallRawOwner(tok.Owner.Raw)
+	if err != nil {
+		return err
+	}
+	script := &Script{}
+	switch owner.Type {
+	case ScriptTypeExchange:
+		err := json.Unmarshal(owner.Identity, script)
+		if err != nil {
+			return errors.Errorf("failed to unmarshal RawOwner as an exchange script")
+		}
+
+		// TODO: does the pre-image match?
```

Comment:
this should be handled by the exchange verifier. the pre-image is part of the signature that is checked against a verifier that encodes the hash.